File: 1019-next-greater-node-in-linked-list/1019-next-greater-node-in-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nextLargerNodes(ListNode head) {
        int n = 0;
        ListNode temp = head;
        while(temp != null) {
            temp = temp.next;
            n++;
        }
        int[] ans = new int[n];
        Stack<Pair<Integer, Integer>>st = new Stack();
        temp = head;
        for(int i=0; i<n; i++) {
            while(!st.isEmpty() && st.peek().getKey() < temp.val) {
                ans[st.peek().getValue()] = temp.val;
                st.pop();
            }
            st.push(new Pair<>(temp.val, i));
            temp = temp.next;
        }
        return ans;
    }
}

File: 1019-next-greater-node-in-linked-list/1019-next-greater-node-in-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nextLargerNodes(ListNode* head) {
        int n = 0;
        ListNode* cur = head;
        while(cur) {
            n++;
            cur = cur->next;
        }
        vector<int>ans(n, 0);
        stack<pair<int, int>>st;
        cur = head;
        for(int i=0; i<n; i++) {
            while(!st.empty() && st.top().first < cur->val) {
                ans[st.top().second] = cur->val;
                st.pop();
            }
            st.push({cur->val, i});
            cur = cur->next;
        }
        return ans;
    }
};